File: Solutions/derived/cartesian1.f90
program cartesian
    use mpi
    implicit none
    integer (kind=4), parameter  :: ndims=2, M=20, N=20

    integer (kind=4) :: i,j,k,ierror,reorder,matsize,localsum,globalsum
    integer (kind=4) :: world_rank, cart_rank
    integer (kind=4) :: world_size, cart_size
    integer (kind=4) :: Comm_cart, prev_rank, next_rank

    integer (kind=4), dimension(ndims) :: dims, periods
    integer (kind=4) :: A(N,M), randvec(N*M)
    real (kind=4) :: rvec(N*M)


    call MPI_Init( ierror )
    if (ierror .NE. 0) then
      write(6,*) ' Cannot start MPI '
      call MPI_Abort(MPI_COMM_WORLD,100,ierror)
    endif
 

    
! Get the world size and rank
    call MPI_Comm_size( MPI_COMM_WORLD, world_size, ierror )
    call MPI_Comm_rank( MPI_COMM_WORLD, world_rank, ierror ) 

    dims = 0
    call MPI_Dims_create( world_size, ndims, dims, ierror );

! Print the dims array to show how it has been changed
    if ( world_rank .EQ. 0 ) then
        write(6,*) 'dims array: ',dims(1), dims(2)
    endif

! Both dimensions will be periodic
    periods = 1
    reorder = 0

! Initialise our new communicator, note the name is informative
! Create it
    call MPI_Cart_create( MPI_COMM_WORLD, ndims, dims, periods, reorder, comm_cart, ierror )
    if (ierror .NE. 0) then
      write(6,*) ' Cannot create Comm_cart '
      call MPI_Abort(MPI_COMM_WORLD,101,ierror)
    endif


! Find the new size and rank
    call MPI_Comm_size( comm_cart, cart_size, ierror )
    call MPI_Comm_rank( comm_cart, cart_rank, ierror )

! Print some information about our rank
    write(6,'(a,i4,a,i4,a,i4,a,i4,a)') 'I am process ',world_rank,' of ', world_size,' on comm_world and ', &
           cart_rank,' of ',cart_size,' on comm_cart.'

   

! Compute the neighbours for each process
    do i=0,ndims-1
       call MPI_Cart_shift( comm_cart, i, 1, prev_rank, next_rank, ierror );
       write(6,'(a,i4,a,i2,a,i4,a,i4)') 'Process ',cart_rank,' on comm_cart. Direction ',i,' prev ', &
              prev_rank,' next ',next_rank
    end do

    call MPI_Finalize(ierror);

end program cartesian

File: ProblemSet2/cartesian1.c
#include <stdlib.h>
#include <mpi.h>
#include <time.h>
#include <math.h>
#include "alloc2d.h"

#define N 20
#define M 20

int main( int argc, char **argv ) {
    MPI_Init( &argc, &argv );

    int i, j;
    int world_rank, cart_rank;
    int world_size, cart_size;
    
    // Get the world size and rank
    MPI_Comm_size( MPI_COMM_WORLD, &world_size );
    MPI_Comm_rank( MPI_COMM_WORLD, &world_rank );

    int ndims = 1;
    int dims[1] = {0};
    // Creates a division of processors on a 1D line
    MPI_Dims_create( world_size, ndims, &dims[0] );
    // Print the dims array to show how it has been changed
    if ( world_rank == 0 ) {
        printf("dims array: %d %d\n", dims[0] );
    }

    // Both dimensions will be periodic
    int periods[1] = { 1 };
    int reorder = 0;
    // Initialise our new communicator, note the name is informative
    MPI_Comm comm_cart;
    // Create a new virtual topology
    MPI_Cart_create( MPI_COMM_WORLD, ndims, dims, periods, reorder, &comm_cart );

    // Find the new size and rank
    MPI_Comm_size( comm_cart, &cart_size );
    MPI_Comm_rank( comm_cart, &cart_rank );

    // Print some information about our rank
    printf( "I am process %d of %d on comm_world and %d of %d on comm_cart.\n", world_rank, world_size, cart_rank, cart_size );

    // Compute the neighbours for each process
    for ( i = 0; i < ndims; i++ ) {
        int prev_rank;
        int next_rank;
		printf("%d\n ", i);
        MPI_Cart_shift( comm_cart, i, 1, &prev_rank, &next_rank );
        printf( "Process %d on comm_cart. Direction %d, prev %d, next %d.\n", cart_rank, i, prev_rank, next_rank );
    }

    MPI_Finalize();
    return 0;
}
